process_scheduling_algorithms/non_preemptive_priority_scheduling.c

```diff
@@ -0,0 +1,407 @@
+    int ct = 0, i, time = 0;
+    int n = LLength(root);
+    float avg, sum = 0;
+    struct node *duproot = NULL;
+    // create a duplicate queue
+    while (ptr != NULL)
+    {
+        insert(&duproot, ptr->ID, ptr->AT, ptr->BT, ptr->priority);
+        ptr = ptr->next;
+    }
+    ptr = duproot;
+    rpt = ptr->next;
+    // sort the queue based on the arrival time and priority
+    while (rpt != NULL)
+    {
+        if (!compare(ptr, rpt))
+        {
+            ptr = rpt;
+        }
+        rpt = rpt->next;
+    }
+    // ptr is the process to be executed first.
+    ct = ptr->AT + ptr->BT;
+    time = ct;
+    sum += ct;
+    // update the completion time, turn around time and waiting time of the
+    // process
+    update(root, ptr->ID, ct, 0, 0);
+    delete (&duproot, ptr->ID);
+    // repeat the process until all the processes are executed
+    for (i = 0; i < n - 1; i++)
+    {
+        ptr = duproot;
+        while (ptr != NULL && ptr->AT > time)
+        {
+            ptr = ptr->next;
+        }
+        rpt = ptr->next;
+        while (rpt != NULL)
+        {
+            if (rpt->AT <= time)
+            {
+                if (rpt->priority < ptr->priority)
+                {
+                    ptr = rpt;
+                }
+            }
+            rpt = rpt->next;
+        }
+        ct += ptr->BT;
+        time += ptr->BT;
+        sum += ct;
+        update(root, ptr->ID, ct, 0, 0);
+        delete (&duproot, ptr->ID);
+    }
+    avg = sum / n;
+    return avg;
+}
+/**
+ * @brief To calculate the average turn around time of all the processes
+ * @param root pointer to the head of the queue
+ * @returns float average turn around time
+ */
+float CalculateTAT(struct node **root)
+{
+    float avg, sum = 0;
+    int n = LLength(root);
+    struct node *ptr = *root;
+    // calculate the completion time if not already calculated
+    if (ptr->CT == 0)
+    {
+        CalculateCT(root);
+    }
+    // calculate the total turn around time of all the processes
+    while (ptr != NULL)
+    {
+        ptr->TAT = ptr->CT - ptr->AT;
+        sum += ptr->TAT;
+        ptr = ptr->next;
+    }
+    avg = sum / n;
+    return avg;
+}
+/**
+ * @brief To calculate the average waiting time of all the processes
+ * @param root pointer to the head of the queue
+ * @returns float average waiting time
+ */
+float CalculateWT(struct node **root)
+{
+    float avg, sum = 0;
+    int n = LLength(root);
+    struct node *ptr = *root;
+    // calculate the completion if not already calculated
+    if (ptr->CT == 0)
+    {
+        CalculateCT(root);
+    }
+    // calculate the total waiting time of all the processes
+    while (ptr != NULL)
+    {
+        ptr->WT = (ptr->TAT - ptr->BT);
+        sum += ptr->WT;
+        ptr = ptr->next;
+    }
+    avg = sum / n;
+    return avg;
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test()
+{
+    // Entered processes
```

Comment:
Test cases should use `assert` to ensure the algorithm works fine and as excepted.